drivers/watchdog/wdt_xilinx_axi.c

```diff
@@ -0,0 +1,234 @@
+		ret = -EPERM;
+		goto out;
+	}
+	if (!data->wdt_started) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	sys_write32(CSR0_WRS | CSR0_WDS, config->base + REG_TWCSR0);
+	sys_write32(0, config->base + REG_TWCSR1);
+	data->wdt_started = false;
+	ret = 0;
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int wdt_xilinx_axi_install_timeout(const struct device *dev,
+					  const struct wdt_timeout_cfg *cfg)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	uint32_t timer_width;
+	bool good_timer_width = false;
+	int ret;
+
+	if (data->timeout_active) {
+		ret = -ENOMEM;
+		goto out;
+	}
+	if (!(cfg->flags & WDT_FLAG_RESET_SOC)) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+	if (cfg->window.min != 0) {
+		ret = -EINVAL;
+		goto out;
+	}
+	for (timer_width = TIMER_WIDTH_MIN; timer_width <= config->timer_width_max; timer_width++) {
+		/* Note: WDT expiry happens after 2 wraps of the timer (first raises an interrupt
+		 * which is not used, second triggers a reset) so add 1 to timer_width.
+		 */
+		const uint64_t expiry_cycles = ((uint64_t)1) << (timer_width + 1);
+		const uint64_t expiry_msec = expiry_cycles * 1000 / config->clock_rate;
+
+		if (expiry_msec >= cfg->window.max) {
+			LOG_INF("Set timer width to %u, actual expiry %u msec", timer_width,
+				(unsigned int)expiry_msec);
+			good_timer_width = true;
+			break;
+		}
+	}
+	if (!good_timer_width) {
+		LOG_ERR("Cannot support timeout value of %u msec", cfg->window.max);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	sys_write32(timer_width, config->base + REG_MWR);
+	data->timeout_active = true;
+	ret = 0;
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int wdt_xilinx_axi_feed(const struct device *dev, int channel_id)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	uint32_t twcsr0 = sys_read32(config->base + REG_TWCSR0);
+	int ret;
+
+	if (channel_id != 0 || !data->timeout_active) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	twcsr0 |= CSR0_WRS | CSR0_WDS;
+	if (data->wdt_started) {
+		twcsr0 |= CSR0_EWDT1;
+	}
+	sys_write32(twcsr0, config->base + REG_TWCSR0);
+	ret = 0;
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int wdt_xilinx_axi_init(const struct device *dev)
+{
+	const struct xilinx_wdt_axi_config *config = dev->config;
+	struct xilinx_wdt_axi_data *data = dev->data;
+
+	data->last_reset_by_wdt = (sys_read32(config->base + REG_TWCSR0) & CSR0_WRS) != 0;
+	if (data->last_reset_by_wdt) {
+		LOG_WRN("Last reset was triggered by WDT");
+		sys_write32(CSR0_WRS, config->base + REG_TWCSR1);
+	}
+	return 0;
+}
+
+int wdt_xilinx_axi_get_last_watchdog_reset(const struct device *dev, bool *last_watchdog_reset)
+{
+	struct xilinx_wdt_axi_data *data = dev->data;
+
+	*last_watchdog_reset = data->last_reset_by_wdt;
+	return 0;
+}
```

Comment:
Reset reason should be exposed through the HWINFO API.

Reply:
It seems that HWINFO API is for managing status arch (or board?) specific registers.
Watchdog driver is better to be independent from HWINFO, I think.

Do we need generic watchdog API to get the reason of watchdog reset or not?

Reply:
The HWINFO API seems like it would make some sense for retrieving the reset reason, but it's also a global thing that seems intended to be for a particular SoC etc. whereas this is a logic core that could potentially be instantiated on various platforms and doesn't have much visibility at that level. So if it exposed a HWINFO API then it could potentially conflict with one already supplied elsewhere. Possibly it could be enabled using a config option or something I suppose?

Reply:
Well, an FPGA with a CPU complex (hard or soft) and a number of IPs is pretty much a SoC, just not an ASIC. I think it would make sense to expose the Xilinx AXI Timebase WDT reset reason using the HWINFO API, but making it optional with a driver-local Kconfig is a good idea. This would allow users to decide override it as needed.

Reply:
Added in optional support for implementing HWINFO API to retrieve reset reason, rather than a driver-private API call.